### PR TITLE
HOSTEDCP-1732: add PDB to sharedIngress router pod

### PR DESCRIPTION
**Pull request by Patryk-Stefanski**
**What this PR does / why we need it**: add PDB to sharedIngress router pod

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes # [HOSTEDCP-1732](https://issues.redhat.com/browse/HOSTEDCP-1732)

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

File: hypershift-operator/controllers/sharedingress/sharedingress_controller.go [hunk 1]
```diff
@@ -198,7 +199,18 @@ func (r *SharedIngressReconciler) reconcileRouter(ctx context.Context, pullSecre
 		}
 	}
 
-	// TODO(alberto): set PDBs.
+	routerDeploymentOwnerRef := config2.OwnerRefFrom(deployment)
+
+	pdb := RouterPodDisruptionBudget()
+	if result, err := r.createOrUpdate(ctx, r.Client, pdb, func() error {
+		ReconcileRouterPodDisruptionBudget(pdb, routerDeploymentOwnerRef)
```
Comment (enxebre) [inline hypershift-operator/controllers/sharedingress/sharedingress_controller.go:206]: Should we also increase the deployment replicas to 2?

Comment (Patryk-Stefanski) [reply hypershift-operator/controllers/sharedingress/sharedingress_controller.go:206]: done

File: hypershift-operator/controllers/sharedingress/router.go [hunk 2]
```diff
@@ -145,6 +144,7 @@ func ReconcileRouterConfiguration(cm *corev1.ConfigMap, config string) error {
 
 func ReconcileRouterDeployment(deployment *appsv1.Deployment, config *corev1.ConfigMap) error {
 	deployment.Spec = appsv1.DeploymentSpec{
+		Replicas: ptr.To(int32(2)),
```
Comment (enxebre) [inline hypershift-operator/controllers/sharedingress/router.go:147]: this deployment should also have affinity preferences to distribute the replicas across Zones/Nodes.
Similar to what we already do with setMultizoneSpread/setNodeSpread for each HCP pods

Comment (Patryk-Stefanski) [reply hypershift-operator/controllers/sharedingress/router.go:151]: done

File: hypershift-operator/controllers/sharedingress/router.go [hunk 3]
```diff
@@ -143,16 +142,21 @@ func ReconcileRouterConfiguration(cm *corev1.ConfigMap, config string) error {
 	return nil
 }
 
-func ReconcileRouterDeployment(deployment *appsv1.Deployment, config *corev1.ConfigMap) error {
+func ReconcileRouterDeployment(deployment *appsv1.Deployment, configMap *corev1.ConfigMap) error {
+	routerDeploymentConfig := config.DeploymentConfig{}
+	routerDeploymentConfig.SetMultizoneSpread(hcpRouterLabels(), false)
+	routerDeploymentConfig.SetNodeSpread(hcpRouterLabels())
```
Comment (enxebre) [inline hypershift-operator/controllers/sharedingress/router.go:148]: node spread uses a required during scheduling anti affinity would will fail the deployment availability for single node case. Let's use only SetMultizoneSpread for now

File: hypershift-operator/controllers/sharedingress/sharedingress_controller.go [hunk 4]
```diff
@@ -4,6 +4,7 @@ import (
 	"context"
 	"fmt"
 	"github.com/openshift/hypershift/cmd/log"
+	config2 "github.com/openshift/hypershift/support/config"
```
Comment (sjenning) [inline hypershift-operator/controllers/sharedingress/sharedingress_controller.go:7]: elsewhere this is `suppconfig`, `supportconfig`, or `hcpconfig`

**Review by sjenning — COMMENTED** [review]
Generally looks good. One nit on import name.

Just to be clear, this is hard coding `replicas: 2` and PDB `maxUnavailable: 1` for the shared ingress deployment created by the HO.

**Comment by Patryk-Stefanski** [comment]
> Just to be clear, this is hard coding `replicas: 2` and PDB `maxUnavailable: 1` for the shared ingress deployment created by the HO.

Yep

File: hypershift-operator/controllers/sharedingress/router.go [hunk 5]
```diff
@@ -143,16 +142,20 @@ func ReconcileRouterConfiguration(cm *corev1.ConfigMap, config string) error {
 	return nil
 }
 
-func ReconcileRouterDeployment(deployment *appsv1.Deployment, config *corev1.ConfigMap) error {
+func ReconcileRouterDeployment(deployment *appsv1.Deployment, configMap *corev1.ConfigMap) error {
```
Comment (bryan-cox) [inline hypershift-operator/controllers/sharedingress/router.go:145]: Can we add a unit test for this?

**Review by bryan-cox — COMMENTED** [review]
The commit messages need the why added (5 in https://hypershift-docs.netlify.app/contribute/#prior-to-submitting-a-pull-request).

File: hypershift-operator/controllers/sharedingress/sharedingress_controller.go [hunk 6]
```diff
@@ -213,7 +214,18 @@ func (r *SharedIngressReconciler) reconcileRouter(ctx context.Context, pullSecre
...
+	routerDeploymentOwnerRef := supportconfig.OwnerRefFrom(deployment)
+
+	pdb := RouterPodDisruptionBudget()
+	if result, err := r.createOrUpdate(ctx, r.Client, pdb, func() error {
+		ReconcileRouterPodDisruptionBudget(pdb, routerDeploymentOwnerRef)
+		return nil
+	}); err != nil {
+		return fmt.Errorf("failed to reconcile etcd pdb: %w", err)
+	} else {
+		log.Log.Info("reconciled etcd pdb", "result", result)
```
Comment (fxierh) [inline hypershift-operator/controllers/sharedingress/sharedingress_controller.go:226]: I believe this should be "reconciled router pdb"

**Pull request merged**